File: problems/06/invert_binary_tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def invert(tree: Optional[TreeNode] = None) -> Optional[TreeNode]:
    """
    Invert a binary tree by swapping the children of each node.

    Args:
        tree (TreeNode): tree to invert. Can be None.
    Returns:
        TreeNode: the inverse of the tree. If the tree was None, then None is returned
    """
    if tree is None:
        return None

    invert(tree.left)
    invert(tree.right)
    tree.left, tree.right = tree.right, tree.left

    return tree


def test_invert_binary_tree():
    # case 1
    first = TreeNode(
        4,
        TreeNode(
            2,
            TreeNode(1),
            TreeNode(3)
        ),
        TreeNode(
            7,
            TreeNode(6),
            TreeNode(9)
        )
    )
    invert(first)
    output = [
        first.val,
        first.left.val, first.right.val,
        first.left.left.val, first.left.right.val,
        first.right.left.val, first.right.right.val,
    ]
    expected = [4, 7, 2, 9, 6, 3, 1]
    assert output == expected

    # case 2
    second = TreeNode(2, TreeNode(1), TreeNode(3))
    invert(second)
    assert [2, 3, 1] == [second.val, second.left.val, second.right.val]

    # case 3
    assert invert() is None


if __name__ == "__main__":
    import pytest
    pytest.main(["-xv", __file__])

File: problems/06/invert_binary_tree.exs
defmodule TreeNode do
  @type t :: %__MODULE__{
          val: integer,
          left: TreeNode.t() | nil,
          right: TreeNode.t() | nil
        }
  defstruct val: 0, left: nil, right: nil

  def invert_tree(nil), do: nil

  def invert_tree(tree) do
    %TreeNode{
      val: tree.val,
      left: invert_tree(tree.right),
      right: invert_tree(tree.left)
    }
  end
end

defmodule TestTreeInvert do
  use ExUnit.Case

  test "leetcode 1" do
    tree = %{
      val: 4,
      left: %{
        val: 2,
        left: %{val: 1, left: nil, right: nil},
        right: %{val: 3, left: nil, right: nil}
      },
      right: %{
        val: 7,
        left: %{val: 6, left: nil, right: nil},
        right: %{val: 9, left: nil, right: nil}
      }
    }

    inverted = TreeNode.invert_tree(tree)
    assert inverted.val == 4
    assert inverted.left.val == 7
    assert inverted.left.left.val == 9
    assert inverted.left.right.val == 6
    assert inverted.right.val == 2
    assert inverted.right.left.val == 3
    assert inverted.right.right.val == 1
  end

  test "leetcode 2" do
    tree = %{
      val: 2,
      left: %{
        val: 1,
        left: nil,
        right: nil
      },
      right: %{
        val: 3,
        left: nil,
        right: nil
      }
    }

    inverted = TreeNode.invert_tree(tree)
    assert inverted.val == 2
    assert inverted.left.val == 3
    assert inverted.right.val == 1
    assert inverted.left.left == nil
    assert inverted.left.right == nil
    assert inverted.right.left == nil
    assert inverted.right.right == nil
  end

  test "leetcode 3" do
    assert TreeNode.invert_tree(nil) == nil
  end
end
